First pass at updated helm/deploy docs (#2638)
rancher/docs: content/k3s/latest/en/helm/_index.md

weight: 42
---

> Helm is the package management tool of choice for Kubernetes. Helm charts provide templating syntax for Kubernetes YAML manifest documents. With Helm we can create configurable deployments instead of just using static files. For more information about creating your own catalog of deployments, check out the docs at [https://helm.sh/docs/intro/quickstart/](https://helm.sh/docs/intro/quickstart/).

Reviewer: I think we may still want to keep this bit around for a while. Some people still play with v1.0.0 and older v1.16.x for some reason. "K3s release _v1.17.0+k3s.1_ added support..."

Reviewer: I see it was moved down below to the upgrading from helm v2 section, but do we need to call this out earlier?

Author: I think this whole section is weird. We kinda start telling people about Helm in general, and point at docs on how to use it standalone (including pointing it at the kubeconfig file) but then switch to talking about using the Helm controller, which doesn't require any of the Helm CLI tools or configuration.

Reviewer 2: Yeah, it's kind of weird. To be clear, the introduction can have another sentence that explicitly says you can deploy Helm charts without the CLI, and most of this page focuses on that.

Author: Thanks! I tweaked the wording a bit.

K3s does not require any special configuration to use with Helm 3. Just be sure you have properly set up your kubeconfig as per the section about [cluster access.](../cluster-access). K3s does include some extra functionality to make deploying both traditional Kubernetes resource manifests and Helm Charts even easier.

This section covers the following topics:

> - [Automatically Deploying Manifests and Helm Charts](#deploying-manifests-and-helm-charts)

Reviewer: Broken anchor suggested fix

```suggestion
- [Automatically Deploying Manifests and Helm Charts](#automatically-deploying-manifests-and-helm-charts)
```

- [Using the Helm CRD](#using-the-helm-crd)
- [Customizing Packaged Components with HelmChartConfig](#customizing-packaged-components-with-helmchartconfig)

> - [Upgrading Helm](#upgrading-helm)

Reviewer: Broken anchor suggested fix

```suggestion
- [Upgrading from Helm v2](#upgrading-from-helm-v2)
```

### Automatically Deploying Manifests and Helm Charts

> Any Kubernetes manifests found in `/var/lib/rancher/k3s/server/manifests` will automatically be deployed to K3s in a manner similar to `kubectl apply`. Manifests deployed in this manner are managed as AddOn custom resources, and can be viewed by running `kubectl get addon -A`. You will find AddOns for packaged components such as CoreDNS, ServiceLB, Traefik, etc. AddOns are created automatically by the deploy controller, and are named based on their filename in the manifests directory.

Reviewer 3: ServiceLB is not included in manifests directory

Author: right you are!

        enabled: true
```

#### HelmChart Field Definitions

| Field | Default | Description | Helm Argument/Flag Equivalent |
|-------|---------|-------------|-------------------------------|
| name | N/A | Helm Chart name | NAME |
| spec.chart | N/A | Helm Chart name in repository, or complete HTTPS URL to chart archive (.tgz) | CHART |
| spec.targetNamespace | default | Helm Chart target namespace | `--namespace` |
| spec.version | N/A | Helm Chart version (when installing from repository) | `--version` |
| spec.repo | N/A | Helm Chart repository URL | `--repo` |
| spec.helmVersion | v3 | Helm version to use (`v2` or `v3`) | N/A |
| spec.bootstrap | False | Set to True if this chart is needed to bootstrap the cluster (Cloud Controller Manager, etc) | N/A |
| spec.set | N/A | Override simple default Chart values. These take precedence over options set via valuesContent. | `--set` / `--set-string` |
| spec.valuesContent | N/A | Override complex default Chart values via YAML file content | `--values` |
| spec.chartContent | N/A | Base64-encoded chart archive (.tgz) | N/A |

### Customizing Packaged Components with HelmChartConfig

> To allow overriding values for packaged components that are deployed as HelmCharts (ie; Traefik), K3s versions `<= v1.19.0+k3s1` support customizing deployments via a HelmChartConfig resources. The HelmChartConfig resource must match the name and namespace of its corresponding HelmChart. The HelmChartConfig allows specifying additional `valuesContent`, which is passed to the `helm` command as an additional value file. Note that value files cannot override options that are customized via the `spec.set` field on the HelmChart.

Reviewer: Nit:

```suggestion
To allow overriding values for packaged components that are deployed as HelmCharts (e.g. Traefik), K3s versions `<= v1.19.0+k3s1` support customizing deployments via a HelmChartConfig resources. The HelmChartConfig resource must match the name and namespace of its corresponding HelmChart. The HelmChartConfig allows specifying additional `valuesContent`, which is passed to the `helm` command as an additional value file. Note that value files cannot override options that are customized via the `spec.set` field on the HelmChart.
```

`(ie; Traefik)` -> `(e.g. Traefik)`

FYI, i.e. ~= "that is to say" think of it like "or to put it another way" or to explain something. Whereas e.g. ~= "for example"

Reviewer: Suggestion: Just say "K3s versions greater than or equal to v1.19.0+k3s1" instead of <=

Author: The <= convention is used in a couple other places. Lets see how it looks if I change em all.

Author: although I had the wrong comparison anyways...

Reviewer: Yeah. It's kind of a nit to use this convention my thought process is just make this more easily readable and friendly. More of an opinion really. I would suggest we shouldn't use this convention anywhere in any of our docs
